### ISSUE TITLE
[BUG]: Random mistaken "Failed to load xxx.bin" toast messages when using JS API

### ISSUE THREAD

**Issue by silidev** [ankidroid/Anki-Android#16510]
### Checked for duplicates?

- [X] This issue is not a duplicate

### Does it also happen in the desktop version?

- [X] This bug does not occur in the latest version of Anki

### What are the steps to reproduce this bug?

Use the JS API a lot. It only happens randomly. Maybe it is possible to reproduce it reliably by calling some harmless API method in a loop.

### Expected behaviour

No error message should be displayed.

### Actual behaviour

Sometimes (no discernible pattern) AnkiDroid will throw a "Card Content Error: Failed to load 'xxx.bin'" warning, despite everything working fine. xxx being a name referring to an element of the JS API, e.g . "cardDue.bin", "cardMod.bin", "ttsSetLanguage". Maybe all strings from the jsApiList in AnkiDroid/src/main/assets/scripts/js-api.js are possible.

### Debug info

```text
AnkiDroid Version = 2.18.0 (464cf290cc77c3f2c562cb03b408b78a13be261b)

Backend Version = 0.1.38-anki24.04.1 (24.04.1 ccd9ca1a8309b80bcb50ddc5d99c7ce63440bce9)

Android Version = 13 (SDK 33)

ProductFlavor = full

Manufacturer = Google

Model = Pixel 4a

Hardware = sunfish

Webview User Agent = Mozilla/5.0 (Linux; Android 13; Pixel 4a Build/TQ3A.230805.001.S1; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/124.0.6367.179 Mobile Safari/537.36

ACRA UUID = 4d6cad38-e805-4913-82fe-99652a272d86

FSRS Enabled = true

Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

I am pretty sure this began with JS-API 0.2 or 0.3.

This is the string: <string name="card_viewer_could_not_find_image">Card Content Error: Failed to load ‘%s’</string> 

This issue is related: https://github.com/ankidroid/Anki-Android/issues/10033

![WhatsApp Image 2024-05-30 at 22 18 58_9a9ce0e9](https://github.com/ankidroid/Anki-Android/assets/16080894/1a768223-7ccc-464d-a8d2-b03b7c5f2dd9)

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by david-allison** [comment]
@silidev Could you provide a sample deck?

**Comment by silidev** [comment]
Yes, I will build one and put it here. Thanks for looking into this. 🙂

David Allison ***@***.***> schrieb am So., 2. Juni 2024,
22:14:

**Comment by silidev** [comment]
Hi David! Unfortunately, I cannot reproduce it in my sample deck. My production collection is not suitable for testing by others. I hope I find time to build my own AnkiDroid from source. Then I will try to fix it myself and test it in my production collection. I guess you best close this issue for now. Thanks again for having a look at this.

**Comment by david-allison** [comment]
I can get a slight reproduction with a `CACHE_MISS` error with `showAnswer` when attaching Chrome Dev tools

Let's leave this open, I know where it needs to be fixed, but I don't know why

@silidev do you have logcat access? I could add some diagnostic logs instead, which can be used to resolve this

**Comment by silidev** [comment]
@david-allison I never used logcat, but I have used the adb command line before and guess that it won't be that difficult, will it?

**Comment by silidev** [comment]
@david-allison By the way: Kind of the deeper reason, why this is a problem at all, is this: the toast message with the error is very low on the screen, hiding the answer buttons. If it would be positioned higher on the screen everything would be just fine. 

Other toast messages are also a somewhat annoying, like "1 card rescheduled". In the current position the toast message is hiding the answer buttons. If it would be a little bit higher, higher than the answer buttons, it would be no problem it all.  I understand that you are interested what is causing the error, but repositioning the toast message display to a higher Y-position would improve the user experience for all types of toast messages.

**Comment by silidev** [comment]
David asked me to open this issue, therefore I think we should keep it
open.

--
Viele Grüße - Helge

github-actions[bot] ***@***.***> schrieb am Di., 4. März
2025, 18:38:

**Comment by silidev** [comment]
still searching for solutions

**Comment by jucor** [comment]
I am getting the same error with the Ultimate Geography deck, randomly complaining of a missing JavaScript file for the first few cards, then going away. 
Using ankidroid 2.22.2

**Comment by david-allison** [comment]
@jucor likely a card template issue in that case, or a file wasn't synced

**Comment by jucor** [comment]
Thanks @david-allison for the follow-up! If the file wasn't downloaded, wouldn't the error keep being shown for all cards?

**Comment by david-allison** [comment]
We stop showing the error after a couple of cards so it doesn't cause a frustrating reviewing experience

EDIT: This `.bin` error is an issue on our side which I haven't triaged, but the typical cause of the error is a bad template/missing file

**Comment by jucor** [comment]
Ah! That makes a lot of sense. I'll try and chase down the missing file.
Might also flag it on the shared deck's GitHub.

**Comment by jucor** [comment]
@david-allison OK, I've checked, and weirdly, the offending file `_ug-jsvectormap.min.css` is not present on Android in the `media.collection` folder, but *neither is it present on the desktop version*. However, the desktop version does not throw an error. I suspect that the issue might be with how the Ultimage Geography deck is packaged, because this file is in their `experimental/` folder. I'll open an issue-there and cross-link.

**Comment by jucor** [comment]
@david-allison I've opened https://github.com/anki-geo/ultimate-geography/issues/700 , will update here if things move there.

**Comment by jucor** [comment]
@david-allison : Fixed upstream :) It was a bug in the Ultimate Geography styling template, referring a now-obsolete file, now fixed! Thanks a lot for the guidance, and even more for your work on AnkiDroid!

**Comment by haripri0109r** [comment]
I’d like to work on this.

**Comment by david-allison** [comment]
@haripri0109r Sure! 

* Can you reproduce it?
* Can you reproduce it with the 'new study screen' setting enabled?
